[Issue Refund] Confirm and submit refund

Ecarrion opened the pull request:
closes #3033 
closes #2769 

# Why

This PR connects the wires and gives the ability to a user to issue a refund! 🎉 

_PS: Handling of previous shipping refunds is not yet supported._

# How
The above is accomplished by:
**Show a confirmation alert before issuing the refund**
Done by showing a simple `AlertViewController` in `RefundConfirmationViewController `

**Submit the refund to the API when the user confirms it.**
Updates `RefundConfirmationViewModel` to dispatch the `RefundStore.createRefund` action.

**Show a progress view while the refund is being created.**
Present an overlay controller from `RefundConfirmationViewController` powered by `InProgressViewController`

**Navigate back to the order details if the refund creation is successful.**
Dismiss the overlay and the modal refund flow after in `RefundConfirmationViewController` after the refund creation finishes

**Show an error notice if the refund creation fails**
If there is an error, show a generic notice and `DDLogError` the actual error.

**Update the order details screen to shows the new refund information.**
Update `RefundsStore` to update the refund's order - condensed refunds array to include the newly created refund.

# Gif
Success | Error
---- | ----
![Refund flow!](https://user-images.githubusercontent.com/562080/99484349-30451400-292e-11eb-8fc4-9b99e1c44ee0.gif) | ![error](https://user-images.githubusercontent.com/562080/99484810-2f60b200-292f-11eb-835f-8f84c0d1ace2.gif)


# Testing Steps
**Success**
- Navigate to a non-refunded order
- Tap on the "Issue Refund" button
- Select some items to refund and tap next
- Tap on the "Refund" button
- See the order detail updated with the new refund information.

**Error**
- Navigate to a non-refunded order
- Tap on the "Issue Refund" button
- Select some items to refund and tap next
- Turn off any internet connection
- Tap on the "Refund" button
- See the error notice.

Update release notes:
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Ecarrion committed:
Add tests to submit refund

Ecarrion committed:
Submit Refund via the RefundConfirmationViewController

Ecarrion committed:
Present Progress view and shows notice with error

Ecarrion committed:
Update CondensedRefund when creating a refund

Ecarrion committed:
Add tests for updating refunds

ctarda commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Issue Refunds/RefundConfirmationViewModel.swift:
```diff
@@ -44,18 +48,34 @@ final class RefundConfirmationViewModel {
     ]
 
-    init(details: Details, currencySettings: CurrencySettings = ServiceLocator.currencySettings) {
+    init(details: Details, actionProcessor: StoresManager = ServiceLocator.stores, currencySettings: CurrencySettings = ServiceLocator.currencySettings) {
```
I like this idiom. We get the best of both worlds. Having global dependencies clearly defined by a global entity and being able to inject dependencies without factories and local DI via the constructor without adding the rigidity that DI tends to add to all designs. Nice.

ctarda replied:
(Sorry, that was just me thinking out loud 😊)

Ecarrion replied:
😄

ctarda reviewed (approved):
I tried to break the flow by going back and forth, assigning items to be refunded, and then updating quantities or changing items and I couldn't 😄 

With an active connection:
1.  Alert shown refund:
<img src="https://user-images.githubusercontent.com/2722505/99490113-2cec6f80-29a4-11eb-9fa0-2a5f0c6009a7.png" width="350"/>

2. After the refund is completed, the UI is updated and the refund is reflected
<img src="https://user-images.githubusercontent.com/2722505/99490174-4988a780-29a4-11eb-80a2-2541eb37d0f7.png" width="350"/>

3. There is an error issuing the refund:
<img src="https://user-images.githubusercontent.com/2722505/99490262-78068280-29a4-11eb-99a2-94b651dafce9.png" width="350"/>

:shipit:

Garance91540 reviewed (commented):
Hi @Ecarrion,

Great work on refunds! Here are the few notes I made:

- **Disabled CTA styling should be updated.** 
I know I've seen a PR on that somewhere, so I know we are working on this. But on the Refund screen, the "Next" CTA should be in disabled state until at list one product has been selected, and right now it seems like we are showing a [pink empty CTA](https://d.pr/i/QSueFG). Here's how they should look in [light](https://d.pr/i/p0Prrz) and [dark](https://d.pr/i/AG1PiU) modes. (If you ever have a doubt on our general components, they can be found in the Figma file "System - iOS Components"

- **Update dialog CTAs styling**
I know you were working from outdated designs so that's on me, but in order to stay align with the styling we use for modals in other part of the app, the "Cancel" and "Refund" CTA  color should be iOS Sytem label / Light / Primary label color https://d.pr/i/vKm0vB

- Success snackbar
If possible let's add the success snackbar too when users successfully refunded products: https://d.pr/i/SEZ05U

Ecarrion committed:
Update alert styling

Ecarrion committed:
Show a success notice when the refunds are created

Ecarrion committed:
Fix minor typo

Ecarrion commented:
Hi @Garance91540 Thanks a lot for the review.

> Disabled CTA styling should be updated.

[An external contributor is working on this](https://github.com/woocommerce/woocommerce-ios/issues/3081). But I'll take care of it if it's not finished by this week.

> Update dialog CTAs styling

Updated
<img width="386" alt="Screen Shot 2020-11-18 at 12 38 53 PM" src="https://user-images.githubusercontent.com/562080/99568307-41793980-299d-11eb-86c0-8e8f59407018.png">

> Success snackbar

Updated (Took the screenshot with a typo, but I have fixed it in code 😇 )
<img width="383" alt="Screen Shot 2020-11-18 at 12 49 09 PM" src="https://user-images.githubusercontent.com/562080/99568354-505fec00-299d-11eb-8103-22c81e39d7d4.png">

Garance91540 reviewed (approved):
👏👏👏
